tools/test-setup.sh

```diff
@@ -8,4 +8,10 @@ set -Eeuxo pipefail
 # non-sudo mode.
 PYTHON=$(command -v python3 python | head -n1)
 
-sudo "$PYTHON" -m pip install -U tox "zipp<0.6.0;python_version=='2.7'"
+$PYTHON -m pip install --user "virtualenv>=20.0.10"
+$PYTHON -m pip install --user "tox>=3.8.0" "zipp<0.6.0;python_version=='2.7'"
```

Comment:
why is zipp needed?

Reply:
There is a bug in older versions of pip where it installs a version incompatible with py27.

BTW, this PR is not ready for review yet, I am using it to test various things. I will remove draft status when I consider it ready.